File: workflow/src/alevin_knee_filter.py
#!/usr/bin/env python3
"""
Identify the knee-point on the alevin barcode rank plot (DeduplicatedReads vs rank)
and write the barcodes above the knee to a file.

For simulated data, pass --sim_barcodes to copy the known cell list directly.
"""

import argparse
import math
import shutil
import sys


def find_knee_threshold(counts_sorted_desc):
    """
    Kneedle algorithm on the log10-log10 barcode rank plot.
    Returns the count value at the knee (inclusive lower bound).
    counts_sorted_desc: list of counts in descending order, already filtered to > 0.
    """
    counts = [c for c in counts_sorted_desc if c > 0]
    if len(counts) < 3:
        return counts[-1] if counts else 0.0

    log_r = [math.log10(i + 1) for i in range(len(counts))]
    log_c = [math.log10(c) for c in counts]

    r_min, r_max = log_r[0], log_r[-1]
    c_min, c_max = log_c[-1], log_c[0]

    if r_max == r_min or c_max == c_min:
        return float(counts[-1])

    x = [(r - r_min) / (r_max - r_min) for r in log_r]
    y = [(c - c_min) / (c_max - c_min) for c in log_c]

    # perpendicular distance from the line (0,1) -> (1,0): x + y = 1
    dists = [abs(xi + yi - 1) / math.sqrt(2) for xi, yi in zip(x, y)]
    knee_idx = dists.index(max(dists))
    return float(counts[knee_idx])


def main():
    parser = argparse.ArgumentParser(
        description='Knee-point barcode filter from alevin featureDump.txt')
    parser.add_argument('--feature_dump', required=True,
                        help='Path to alevin featureDump.txt')
    parser.add_argument('--output', required=True,
                        help='Output path for filtered barcodes (one per line)')
    parser.add_argument('--sim_barcodes', default='',
                        help='Simulation mode: copy this file to output, skipping knee detection')
    args = parser.parse_args()

    if args.sim_barcodes:
        shutil.copy(args.sim_barcodes, args.output)
        print(f'simulation mode: copied {args.sim_barcodes} to {args.output}',
              file=sys.stderr)
        return

    barcodes = []
    dedup_reads = []
    with open(args.feature_dump) as fh:
        next(fh)  # skip header
        for line in fh:
            parts = line.rstrip('\n').split('\t')
            barcodes.append(parts[0])
            dedup_reads.append(float(parts[3]))

    counts_sorted = sorted(dedup_reads, reverse=True)
    threshold = find_knee_threshold(counts_sorted)

    kept = [b for b, c in zip(barcodes, dedup_reads) if c >= threshold]
    with open(args.output, 'w') as out:
        for b in kept:
            out.write(b + '\n')

    print(
        f'knee threshold: {threshold:.1f} deduplicated reads, '
        f'kept {len(kept)} / {len(barcodes)} barcodes',
        file=sys.stderr
    )


if __name__ == '__main__':
    main()

File: workflow/src/alevin_knee_filter.R
#!/usr/bin/env Rscript
## Filter alevin barcodes using DropletUtils barcodeRanks.
## Non-sketch mode: reads DeduplicatedReads from featureDump.txt.
## Sketch mode: reads per-cell UMI totals from the alevin-fry MTX output.
## Usage: Rscript alevin_knee_filter.R \
##            (--feature_dump <path> | --fry_quant_dir <path>) \
##            --output <path> \
##            [--cell_filtering <native|emptydrops|none>] \
##            [--sim_barcodes <path>]
## cell_filtering none keeps every observed barcode (no knee filter).

suppressPackageStartupMessages({
    library(DropletUtils)
    library(Matrix)
    library(data.table)
})

args = commandArgs(trailingOnly = TRUE)

parse_args = function(argv) {
    out = list(feature_dump = "", fry_quant_dir = "", output = NULL, sim_barcodes = "",
               cell_filtering = "native")
    i = 1
    while (i <= length(argv)) {
        if (argv[i] == "--feature_dump")   { out$feature_dump = argv[i + 1];   i = i + 2 }
        else if (argv[i] == "--fry_quant_dir") { out$fry_quant_dir = argv[i + 1]; i = i + 2 }
        else if (argv[i] == "--output")    { out$output = argv[i + 1];         i = i + 2 }
        else if (argv[i] == "--cell_filtering") { out$cell_filtering = argv[i + 1]; i = i + 2 }
        else if (argv[i] == "--sim_barcodes") { out$sim_barcodes = argv[i + 1]; i = i + 2 }
        else i = i + 1
    }
    out
}

opts = parse_args(args)

if (nchar(opts$sim_barcodes) > 0) {
    file.copy(opts$sim_barcodes, opts$output, overwrite = TRUE)
    message("simulation mode: copied ", opts$sim_barcodes, " to ", opts$output)
    quit(save = "no", status = 0)
}

if (nchar(opts$fry_quant_dir) > 0) {
    fry_alevin = file.path(opts$fry_quant_dir, "alevin")
    mtx_f = file.path(fry_alevin, "quants_mat.mtx")
    rows_f = file.path(fry_alevin, "quants_mat_rows.txt")

    mat = readMM(mtx_f)
    barcodes = readLines(rows_f)
    umi_totals = rowSums(mat)
} else {
    fd = fread(opts$feature_dump, sep = "\t", header = TRUE)
    barcodes = fd[[1]]
    umi_totals = as.numeric(fd[["DeduplicatedReads"]])
}

if (opts$cell_filtering == "none") {
    writeLines(barcodes, opts$output)
    message(sprintf("cell_filtering none: kept all %d barcodes (no knee filter)",
                    length(barcodes)))
    quit(save = "no", status = 0)
}

br = barcodeRanks(matrix(umi_totals, nrow = 1, dimnames = list(NULL, barcodes)))
knee_threshold = metadata(br)$knee

if (is.na(knee_threshold)) {
    knee_threshold = metadata(br)$inflection
    warning(sprintf("barcodeRanks knee is NA; falling back to inflection point: %g",
                    knee_threshold))
}

if (is.na(knee_threshold)) {
    warning("both knee and inflection are NA; keeping all barcodes")
    keep = rep(TRUE, length(barcodes))
} else {
    keep = umi_totals >= knee_threshold
}

writeLines(barcodes[keep], opts$output)

message(sprintf(
    "barcodeRanks knee: kept %d / %d barcodes (threshold: %g)",
    sum(keep), length(barcodes), knee_threshold
))
